src/main/java/com/puppycrawl/tools/checkstyle/SarifLogger.java

```diff
@@ -156,6 +176,58 @@ public SarifLogger(
             readResource("/com/puppycrawl/tools/checkstyle/sarif/ResultFileOnly.template");
         resultErrorOnly =
             readResource("/com/puppycrawl/tools/checkstyle/sarif/ResultErrorOnly.template");
+        rule = readResource("/com/puppycrawl/tools/checkstyle/sarif/Rule.template");
+    }
+
+    /**
+     * Loads all available module metadata from XML files.
+     */
+    private void loadModuleMetadata() {
+        final List<ModuleDetails> allModules =
+                XmlMetaReader.readAllModulesIncludingThirdPartyIfAny();
+        allModuleMetadata = new LinkedHashMap<>(allModules.size());
+        for (ModuleDetails module : allModules) {
+            allModuleMetadata.put(module.getFullQualifiedName(), module);
+        }
+    }
+
+    /**
+     * Finds and caches module metadata for a given rule ID.
+     *
+     * @param ruleId the rule ID (violation key)
+     */
+    private void cacheRuleMetadata(String ruleId) {
+        if (!ruleMetadata.containsKey(ruleId)) {
+            for (ModuleDetails module : allModuleMetadata.values()) {
+                if (module.getViolationMessageKeys().contains(ruleId)) {
+                    ruleMetadata.put(ruleId, module);
+                    break;
```

Comment:
```suggestion
                    return;
```
its complete, lets avoid any further.